cylc/uiserver/handlers.py

```diff
@@ -145,9 +150,58 @@ def prepare(self):
         super().prepare()
 
 
+class SubscriptionHandler(websocket.WebSocketHandler):
+    subscription_server = None
+    resolvers = None
+    queue = Queue(100)
+
+    def initialize(self, sub_server, resolvers):
+        self.subscription_server = sub_server
+        self.resolvers = resolvers
+
+    def select_subprotocol(self, subprotocols):
+        return GRAPHQL_WS
+
+    def open(self, *args, **kwargs):
+        IOLoop.current().spawn_callback(self.subscription_server.handle, self,
+                                        self.context)
+
+    async def on_message(self, message):
+        await self.queue.put(message)
+
+    async def recv(self):
+        return await self.queue.get()
+
+    def check_origin(self, origin: str) -> bool:
+        return True
+
+    @property
+    def context(self):
+        wider_context = {
+            'request': self.request,
+            'resolvers': self.resolvers,
+        }
+        return wider_context
```

Comment:
This was necessary but took me a while to understand why. Thanks @dwsutherland for saving me a lot of time! Note for others: the schema and base resolvers are defined in `cylc-flow`. But in `cylc-uiserver` we **use** the schema, but **extend** the resolvers in cylc-uiserver (I believe it's to augment the resolvers with extra data/functionality). So we need to pass the resolvers from cylc-uiserver to cylc-flow... so that when the schema resolves it it uses the right resolvers :grimacing: (feel free to correct me if I said something wrong @dwsutherland)

Reply:
> I believe it's to augment the resolvers with extra data/functionality

Yes, the mutations resolvers are different (at the moment) but query resolver's identical.. `BaseResolvers` is extended in both `cylc-flow` and `cylc-uiserver` .. just in different ways..

> So we need to pass the resolvers from cylc-uiserver to cylc-flow... so that when the schema resolves it it uses the right resolvers 😬 (feel free to correct me if I said something wrong @dwsutherland)

Not sure I follow here.. WRT code, nothings passed from `cylc-uiserver` to `cylc-flow` (only the other way round.. The GraphQL end points at UIS and WS are completely independent of each other, except for Mutations where the UIS passes the request doc directly to the respective WS GraphQLs..
The schema resolves from the same named methods at both UIS & WS, this is to keep the schema identical for both end-points.. 
The main purposes are free upgrades of `cylc-uiserver` when `cylc-flow` is upgraded, and less code to maintain. (keeps the two in sync).. `BaseResolvers` allows us to do the same with what resolvers we are capable of.

Reply:
>Not sure I follow here.. WRT code, nothings passed from cylc-uiserver to cylc-flow (only the other way round)

That's what I am not understanding. Probably I am missing something here. But bear with me a bit more. Let me use some code from `master` to try to explain again what I was trying to say.

- This is the Tornado handler for `/graphql`: https://github.com/cylc/cylc-uiserver/blob/453688c023e39a030c987fd9b4fb172852140fe5/cylc/uiserver/main.py#L109-L111
- It passes a `resolvers=` parameter to the Handler.
- That `resolvers` will be set in the Handler `initialize` method: https://github.com/cylc/cylc-uiserver/blob/453688c023e39a030c987fd9b4fb172852140fe5/cylc/uiserver/handlers.py#L132
- Later, the request will have a `.context` property defined here (still in `cylc-uiserver`): https://github.com/cylc/cylc-uiserver/blob/453688c023e39a030c987fd9b4fb172852140fe5/cylc/uiserver/handlers.py#L135-L141

From what I understood about the design in place, `cylc-flow` has the schema that is used by Cylc UI Server. So Cylc UI Server is independent of Cylc Flow, and just a consumer of its schema. I like this approach.

But now here's the part I found strange. In Cylc Flow, the `schema.py` receives an `info` object, and uses the `context` to retrieve the resolvers: https://github.com/cylc/cylc-flow/blob/9ff5b7719fe924a9a2c8c1b3446faaaf47a4c5b2/cylc/flow/network/schema.py#L275

That resolver is not `cylc.flow.network.resolvers`, but `cylc.uiserver.resolvers`. **So Cylc UI Server has a dependency on the schema of Cylc Flow; and Cylc Flow has a dependency on the resolvers of Cylc UI Server**.

From what I understood, if I made a change in the resolvers of Cylc UI Server that impacted the schema in Cylc Flow, we could end up having to patch/release Cylc Flow. Which is going in the wrong dependency direction? (i.e. I thought Cylc UI Server would be a consumer of Cylc Flow, not the other way around)

But again, I could be missing something here, and saying something wrong - so sorry in case it's just gibberish and also for the long reply :+1:

Reply:
> That resolver is not cylc.flow.network.resolvers, but cylc.uiserver.resolvers. So Cylc UI Server has a dependency on the schema of Cylc Flow; and Cylc Flow has a dependency on the resolvers of Cylc UI Server.

Here's the missing pieces:
https://github.com/cylc/cylc-flow/blob/9ff5b7719fe924a9a2c8c1b3446faaaf47a4c5b2/cylc/flow/network/server.py#L36
Instantiated here:
https://github.com/cylc/cylc-flow/blob/9ff5b7719fe924a9a2c8c1b3446faaaf47a4c5b2/cylc/flow/network/server.py#L268-L270
And passed in here:
https://github.com/cylc/cylc-flow/blob/9ff5b7719fe924a9a2c8c1b3446faaaf47a4c5b2/cylc/flow/network/server.py#L342-L350

So nothing from `cylc-uiserver` is imported or passed to `cylc-flow`

Reply:
AaaaaaaaaahhhhhhhH! Now that makes sense! hahaha. Thanks @dwsutherland !